src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,172 @@
+import * as WebSocket from 'ws';
+import { EventEmitter } from 'events';
+
+import { sbvrUtils } from '@resin/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { requestAsync } from '../../request';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = sbvrUtils;
+
+const VERSION = 1;
+
+interface LokiStream {
+	stream: {
+		device_id: string;
+		service_id: string;
+		is_system: string;
+		is_std_err: string;
+	};
+	values: Array<[string, string]>;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private webSockets: Map<string, WebSocket>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.webSockets = new Map();
+	}
+
+	public get available(): boolean {
+		// TODO: check /ready endpoint?
+		return true;
+	}
+
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const [response, body] = await requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/query_range`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
```

Comment:
Am I right in thinking a device will lose all of its logs if it's transferred between applications?

Reply:
Correct. Queries will only return a device's logs relative to its current application. I will bring up this issue in the check in.